feat: add notifications inbox db

defelmnq opened the pull request:
This PR is linked [to the following issue](https://github.com/coder/internal/issues/334). 

The objective is to create the DB layer and migration for the new `Coder Inbox`.

defelmnq committed:
work on db queries

defelmnq committed:
work on dbauthz

defelmnq committed:
work on dbauthz

defelmnq committed:
fmt and lint

defelmnq committed:
change queries to match targets

defelmnq committed:
test dbauthz

defelmnq committed:
move notification migration

defelmnq committed:
make gen

defelmnq committed:
Merge remote-tracking branch 'origin/main' into notif-inbox/int-334

defelmnq committed:
rename inbox notifications

defelmnq committed:
rename inbox notifications

fix rbac roles

defelmnq committed:
Improve authz

defelmnq committed:
add references in db

defelmnq committed:
fix test icon

defelmnq committed:
improve foreign keys

defelmnq committed:
add back querier file

defelmnq committed:
improve uuid in seeding

defelmnq committed:
improve dbauthz testing

dannykopping commented on coderd/database/queries/notificationsinbox.sql:
```diff
@@ -0,0 +1,38 @@
+SELECT * FROM inbox_notifications WHERE user_id = $1 ORDER BY created_at DESC;
+
+-- name: FetchInboxNotificationsByUserIDFilteredByTemplatesAndTargets :many
+SELECT * FROM inbox_notifications WHERE user_id = @user_id AND template_id = ANY(@templates::UUID[]) AND targets @> COALESCE(@targets, ARRAY[]::UUID[]) ORDER BY created_at DESC;
```
I had to look this up.

https://www.postgresql.org/docs/current/functions-array.html

`@>` implies that _every_ element in `COALESCE(@targets, ARRAY[]::UUID[])` must appear in `targets`. Seems reasonable since the filtering will be ANDs not ORs.

In any case, it would be useful to document the behaviour that callers can expect from this query, and why you chose this specific operator.

dannykopping commented on coderd/database/queries/notificationsinbox.sql:
```diff
@@ -0,0 +1,38 @@
+-- name: FetchUnreadInboxNotificationsByUserIDFilteredByTemplatesAndTargets :many
+SELECT * FROM inbox_notifications WHERE user_id = @user_id AND template_id = ANY(@templates::UUID[]) AND targets @> COALESCE(@targets, ARRAY[]::UUID[]) AND read_at IS NULL ORDER BY created_at DESC;
+
+-- name: GetInboxNotificationByID :one
```
Nit:

`Get` or `Fetch`, pick one please.

```bash
$ egrep ') \bGet' coderd/database/queries.sql.go  | wc -l
189
$ egrep ') \bFetch' coderd/database/queries.sql.go  | wc -l
3
```

dannykopping commented on coderd/database/queries/notificationsinbox.sql:
```diff
@@ -0,0 +1,38 @@
+VALUES
+    ($1, $2, $3, $4, $5, $6, $7, $8, $9) RETURNING *;
+
+-- name: SetInboxNotificationAsRead :exec
```
Nit:

```bash
$ egrep ') \bSet' coderd/database/queries.sql.go | wc -l
0
$ egrep ') \bUpdate' coderd/database/queries.sql.go | wc -l
66
```

defelmnq replied:
Renamed to stick with `Update`

dannykopping commented on coderd/database/migrations/000296_notifications_inbox.up.sql:
```diff
@@ -0,0 +1,15 @@
+CREATE TABLE inbox_notifications (
```
Nit: format this file please.

dannykopping commented on coderd/database/migrations/000296_notifications_inbox.up.sql:
```diff
@@ -0,0 +1,15 @@
+	title			TEXT 						NOT NULL,
+	content 		TEXT 						NOT NULL,
+	icon 			TEXT 						NOT NULL,
+	actions 		JSONB 						NOT NULL,
```
Given that you have `title` and `content`, I assume you're going to be accepting the fully rendered content (probably Markdown) of the notification in these fields. That means that CTAs will already be included in `content`, I think. What role will `actions` play here?

defelmnq replied:
The idea was to generate the title, content and actions using a `Dispatcher` method [similar to SMTP](https://github.com/coder/coder/blob/main/coderd/notifications/dispatch/smtp.go#L63).
the `payload` field contains everything to have these three elements.

The reason why I had in mind to keep the actions separated from the body is that , based on the client (VSCode for example) we will not be able to handle the actions the same way. If we integrate the actions in the body using markdown it will not be compatible with everything. wdyt ?

dannykopping commented on coderd/database/migrations/000296_notifications_inbox.up.sql:
```diff
@@ -0,0 +1,15 @@
+	created_at 		TIMESTAMP WITH TIME ZONE 	NOT NULL DEFAULT NOW()
+);
+
+CREATE INDEX idx_inbox_notifications_user_id_read_at ON inbox_notifications(user_id, read_at);
```
👍 good idea

dannykopping commented on coderd/rbac/roles_test.go:
```diff
@@ -354,6 +354,17 @@ func TestRolePermissions(t *testing.T) {
+			},
+			Resource: rbac.ResourceInboxNotification.WithID(uuid.New()).InOrg(orgID).WithOwner(currentUser.String()),
+			AuthorizeMap: map[bool][]hasAuthSubjects{
+				true:  {owner, orgMemberMe, orgAdmin},
```
Why does `orgMemberMe` have privileges to `policy.ActionCreate` inbox notifications?

defelmnq replied:
From my RBAC understanding `orgMemberMe` is the current user. 
Associated [to this part](https://github.com/coder/coder/blob/main/coderd/rbac/roles_test.go#L113C69-L113C80) on the which we use the `currentUser` id.

In the definition of the ResourceInboxNotification, I assign the `currentUser` to `InboxNotification` using the `WithOwner(currentUser.String())` method.

TL;DR - `orgMemberMe` is the owner of the resource in the test - and the owner of the resource should be able to access it.

johnstcn replied:
I think we might want `memberMe` in here as well?

dannykopping replied:
@defelmnq  I understand why the current user should be able to _read_ their own messages, but why should they be allowed to _create_ new messages? That should be a system privilege.

dannykopping commented on coderd/database/queries/notificationsinbox.sql:
```diff
@@ -0,0 +1,38 @@
+-- name: FetchUnreadInboxNotificationsByUserID :many
+SELECT * FROM inbox_notifications WHERE user_id = $1 AND read_at IS NULL ORDER BY created_at DESC;
+
+-- name: FetchInboxNotificationsByUserID :many
+SELECT * FROM inbox_notifications WHERE user_id = $1 ORDER BY created_at DESC;
```
These queries are identical except for `read_at IS NULL`; rather parameterise `FetchInboxNotificationsByUserID` so you can explicitly filter for read/unread, or neither.

`FetchUnreadInboxNotificationsByUserIDFilteredByTemplatesAndTargets` could be refactored out according to the same pattern, too.

defelmnq replied:
As I tried to iterate here and from what I found on the internet - I can't find a way to achieve that with sqlc.

sqlc, with narg, allow to have null parameters and optional values, with default values if we want - here the objective is a bit different as we want to completely change the query and the `WHERE` clause based on if a parameter is present or not.

If you have more idea, I could check.

defelmnq committed:
format migration

defelmnq committed:
format migration

defelmnq committed:
format migration

defelmnq committed:
rename functions

defelmnq committed:
add back index

defelmnq committed:
fix query name missing

defelmnq committed:
regen file

defelmnq committed:
add count queries

defelmnq committed:
improve queries to add pagination

defelmnq committed:
fix dbauthz tests

defelmnq committed:
fix sqlc queries

defelmnq committed:
improve dbauthz testing

defelmnq commented:
Update on the PR with some changes : 

- Naming for the DB methods has been changed to fit with the existing ones.
- Added a new method `CountUnreadNotificationsByUserID` which was missing
- Added pagination to list methods

About the pagination - because of the way notifications can be pushed (in real-time , potentially a lot of notifs quickly) - I decided to paginate using IDs instead of an offset which would be hard to maintain.

johnstcn commented on coderd/database/queries/notificationsinbox.sql:
```diff
@@ -0,0 +1,71 @@
+-- name: GetInboxNotificationByID :one
+SELECT * FROM inbox_notifications WHERE id = $1;
+
+-- name: CountUnreadInboxNotificationsByUserID :one
+SELECT COUNT(*) FROM inbox_notifications WHERE user_id = $1 AND read_at IS NULL;
```
Where do we plan to use this query? If we're just displaying the unread notifications count in the UI, would it suffice to just call `len(GetUnreadInboxNotificationsByUserID()`? My assumption is that the UI will be fetching unread notifications anyway.

defelmnq replied:
This query will be called in two cases : 

- An endpoint to get the number of unread notifications (without any notifications - just the count)
- Whenever you want to list the number of notifications (unread or not, filtered or not) - we'll include the `unread_count`

johnstcn commented on coderd/database/queries/notificationsinbox.sql:
```diff
@@ -0,0 +1,71 @@
+-- name: GetUnreadInboxNotificationsByUserIDFilteredByTemplatesAndTargets :many
+-- param user_id: The user ID
+-- param templates: The template IDs to filter by - the template_id = ANY(@templates::UUID[]) condition checks if the template_id is in the @templates array
+-- param targets: The target IDs to filter by - the targets @> COALESCE(@targets, ARRAY[]::UUID[]) condition checks if the targets array (from the DB) contains all the elements in the @targets array
```
As a caller, it's not really clear to me what kind of UUID a target should be without looking at example usages. Could this be reworded slightly to be clearer?

johnstcn commented on coderd/database/migrations/000296_notifications_inbox.up.sql:
Obligatory reminder to double-check migration number before merging!

johnstcn reviewed (approved):
No blocking comments, some suggestions below.

defelmnq committed:
improve sql queries

defelmnq committed:
add type for read status

defelmnq committed:
iterate on testing dbauthz

defelmnq committed:
change zero value of timestamp

defelmnq committed:
add returns check in tests

defelmnq commented on coderd/database/types.go:
```diff
@@ -35,6 +35,40 @@ type NotificationsSettings struct {
 	NotifierPaused bool      `db:"notifier_paused" json:"notifier_paused"`
 }
 
+type InboxNotificationReadStatus string
```
@dannykopping this is a new part of the PR that you'd probably want to check , related to our discussion and the new version of SQL queries.

defelmnq committed:
Merge remote-tracking branch 'origin/main' into notif-inbox/int-334

defelmnq committed:
rename migration

dannykopping commented on coderd/database/types.go:
```diff
@@ -35,6 +35,40 @@ type NotificationsSettings struct {
+type InboxNotificationReadStatus string
+
+const (
+	InboxNotificationReadStatusRead   InboxNotificationReadStatus = "READ"
```
Rather create a SQL type for this to get automatic codegen
Nit: make this lowercase to match other types' values

dannykopping commented on coderd/database/queries/notificationsinbox.sql:
```diff
@@ -0,0 +1,51 @@
+	ORDER BY created_at DESC
+	LIMIT (COALESCE(NULLIF(@limit_opt :: INT, 0), 25));
+
+-- name: GetInboxNotificationsByUserIDFilteredByTemplatesAndTargets :many
```
Nit: for future-proofing, rather drop the specificity here in case we add more filter options later:

```suggestion
-- name: GetFilteredInboxNotificationsByUserID :many
```

defelmnq replied:
Renamed ✅

dannykopping commented on coderd/database/queries/notificationsinbox.sql:
```diff
@@ -0,0 +1,51 @@
+
+-- name: GetInboxNotificationsByUserIDFilteredByTemplatesAndTargets :many
+-- Fetches inbox notifications for a user filtered by templates and targets
+-- param user_id: The user ID
```
If you're going to document params, best to include all of them.

defelmnq committed:
improve queries comments

defelmnq committed:
cange type for read status

dannykopping reviewed (approved):
LGTM 👍

defelmnq committed:
merge
